winch/codegen/src/isa/x64/masm.rs

```diff
@@ -916,6 +916,137 @@ impl Masm for MacroAssembler {
         }
     }
 
+    fn truncate(
+        &mut self,
+        context: &mut CodeGenContext,
+        src_size: OperandSize,
+        dst_size: OperandSize,
+        kind: TruncateKind,
+    ) {
```

Comment:
I'll add this comment here, but it applies to all the new methods which take a `&mut CodeGenContext` as a parameter. 

I'm trying to understand why passing the `&mut CodeGenContext` is needed, versus passing the already resolved register operands. The general convention that I'm trying to keep at the MacroAssembler level is that the approach of passing the `CodeGenContext` should be used as a last resort, more concretely when the lowering of a particular instruction is not generic enough, and needs resolution of ISA-specific details, like for example in the case of the `div` instruction, which in x64 requires very specific registers to be used. Looking at the conversion and truncation methods, it doesn't look like there's anything ISA specific here and that most of the logic can be moved to the visitor, which would be my preference. Another advantage of such approach is that by keeping the MacroAssembler layer as thin as possible, there's less repetition that will be needed when working on other backends.

Reply:
Let me know if there's anything that I'm missing here.

Reply:
Yeah I can move the register part of the logic into the visitor. I wasn't entirely sure what your preference would be so I went with one approach but agreed that since it's not ISA-specific, having in the visitor makes more sense.

Reply:
Actually for `truncate`, I need an additional temporary XMM register but only when performing an unsigned truncation. I could change how we model truncation so we have a `signed_truncate` and an `unsigned_truncate` at the MacroAssembler level where the `unsigned_truncate` method would take an additional temporary XMM parameter.

Reply:
The same pattern holds for `convert` where I need an additional temporary GPR but only when performing an unsigned `convert`. Again I could split that method at the MacroAssembler level into a `signed_convert` and an `unsigned_convert` where the `unsigned_convert` takes an additional temporary GPR as a parameter.

Reply:
Yeah, I didn't mention it explicitly on my original comment but I think that splitting those operations (<signed|unsigned>_convert) and perform the dispatch from the visitor makes sense to me.

Reply:
Sorry for rambling on this, it's been a few weeks since I worked on this so I'm just remembering I did this as I go through the exercise again. I don't know if we would need the extra temporary XMM register or GPR for other ISAs when performing unsigned truncation or conversion. Maybe that's just an x86 detail? But I haven't tried making an AArch64 implementation yet so maybe it would also be necessary on that ISA. So does it make sense to bake in needing additional temporary registers for unsigned truncation and conversion at the MacroAssembler trait level given it _may_ just be an x86 requirement?

Reply:
I think it makes sense to do that. When implementing the aarch64 backend we'll know for sure and we can adjust then. If an extra register is the issue, there are other options to deal with that IMO, like for example passing an smallvec of registers to work with, in which each backend defines how many are needed, which could be empty if no extra registers are needed.